e2e-tests/src/test/adder.rs

```diff
@@ -19,26 +18,22 @@ pub async fn adder() -> Result<()> {
 
     let (conn, _authority, account) = basic_test_context(config).await?;
 
-    let contract = Arc::new(AdderInstance::new(
+    let contract = AdderInstance::new(
         &config.test_case_params.adder,
         &config.test_case_params.adder_metadata,
-    )?);
-
-    let listen_conn = conn.clone();
-    let listen_contract = contract.clone();
-    let (tx, mut rx) = unbounded();
-    let listen = || async move {
-        listen_contract_events(&listen_conn, &[listen_contract.as_ref().into()], tx).await?;
-        <Result<(), anyhow::Error>>::Ok(())
-    };
-    let join = tokio::spawn(listen());
+    )?;
 
     let increment = 10;
     let before = contract.get(&conn).await?;
 
-    contract.add(&account.sign(&conn), increment).await?;
+    let tx_info = contract.add(&account.sign(&conn), increment).await?;
+    let events = get_contract_events(&conn, &contract.contract, tx_info).await?;
+    let event = match events.len() {
+        0 => return Err(anyhow!("No event received")),
+        1 => events[0].clone(),
+        _ => return Err(anyhow!("Too many events received")),
+    };
```

Comment:
Tip: You can extract element while pattern matching on the array
```rust
match &*events {
  &[event] => event,
 _ => return Err(...)
}
```

Use or not :)

Reply:
The message could have been `"expected single event but got {:?}, events"` but it's your call :)

Reply:
I agree with @deuszx for the record - less code = better

Reply:
simplified